build/source/engine/read_force.f90

```diff
@@ -417,49 +214,313 @@ subroutine read_force(istep,iFile,iRead,ncid,time_data,forcStruct,err,message)
+ dataJulDay = varTime(1)/forcFileInfo(iFile)%convTime2Days + refJulday_data
+ if(abs(currentJulday - dataJulDay) > verySmall)then
+  write(message,'(a,f18.8,a,f18.8)') trim(message)//'date for time step: ',dataJulDay,' differs from the expected date: ',currentJulDay
+  err=40; return
+ end if
+
+ ! convert julian day to time vector
+ ! NOTE: use small offset to force ih=0 at the start of the day
+ call compcalday(dataJulDay+smallOffset,         & ! input  = julian day
+                 time_data(iLookTIME%iyyy),      & ! output = year
+                 time_data(iLookTIME%im),        & ! output = month
+                 time_data(iLookTIME%id),        & ! output = day
+                 time_data(iLookTIME%ih),        & ! output = hour
+                 time_data(iLookTIME%imin),dsec, & ! output = minute/second
+                 err,cmessage)                     ! output = error control
+ if(err/=0)then; message=trim(message)//trim(cmessage); return; end if
+
+ ! check to see if any of the time data is missing
+ if(any(time_data(:)==integerMissing))then
+  do iline=1,size(time_data)
+   if(time_data(iline)==integerMissing)then; err=40; message=trim(message)//"variableMissing[var='"//trim(time_meta(iline)%varname)//"']"; return; end if
+  end do
+ end if
+
+ ! initialize flags for forcing data
+ checkForce(:) = .false.
+ checkForce(iLookFORCE%time) = .true.  ! time is handled separately
+
+ ! loop through forcing data variables
+ do iNC=1,forcFileInfo(iFile)%nVars
+
+  ! check variable is desired
+  if(forcFileInfo(iFile)%var_ix(iNC)==integerMissing) cycle
+
+  ! get index in forcing structure
+  iVar = forcFileInfo(iFile)%var_ix(iNC)
+  checkForce(iVar) = .true.
+
+  ! read forcing data for all HRUs
+  if(simultaneousRead)then
+   err=nf90_get_var(ncid,forcFileInfo(iFile)%data_id(ivar),dataVec,start=(/ixHRUfile_min,iRead/),count=(/nHRUlocal,1/))
+   if(err/=nf90_noerr)then; message=trim(message)//'problem reading forcing data: '//trim(varName)//'/'//trim(nf90_strerror(err)); return; endif
+  endif
+
+  ! loop through GRUs and HRUs
+  do iGRU=1,size(gru_struc)
+   do iHRU=1,gru_struc(iGRU)%hruCount
+
+    ! define global HRU
+    iHRU_global = gru_struc(iGRU)%hruInfo(iHRU)%hru_nc
+    iHRU_local  = (iHRU_global - ixHRUfile_min)+1
+    !print*, 'iGRU, iHRU, iHRU_global, iHRU_local = ', iGRU, iHRU, iHRU_global, iHRU_local
+
+    ! read forcing data for a single HRU
+    if(.not.simultaneousRead)then
+     err=nf90_get_var(ncid,forcFileInfo(iFile)%data_id(ivar),dataVal,start=(/iHRU_global,iRead/))
+     if(err/=nf90_noerr)then; message=trim(message)//'problem reading forcing data: '//trim(varName)//'/'//trim(nf90_strerror(err)); return; endif
+    endif
+
+    ! check the number of HRUs
+    if(iHRU_global > nHRUfile)then
+     message=trim(message)//'HRU index exceeds the number of HRUs in the forcing data file'
+     err=20; return
+    endif
+
+    ! get individual data value
+    if(simultaneousRead) dataVal(1) = dataVec(iHRU_local)
```

Comment:
Isn't this to get the forcing data for all HRUs? In that case: fix comment.

Reply:
If this is to get the data for all HRUs, shouldn;t this be outside the HRU loop, i.e.:

```Fortran
do iGRU=1,size(gru_struc)
  if(simultaneousRead) ..... 
  else !loop over the HRUs and do individual reads
    do iHRU=1,gru_struc(iGRU)%hruCount .....
    end do
  end if
end do
```

Reply:
This is still only getting one time step per read - is that correct?

Reply:
This is getting one time step per read. We get all HRUs at once if `simultaneousRead==.true.`

Reply:
The code on line 496: This extracts an individual data value from the data vector, so the comment is correct.

The logic is as follows:

```fortran
  ! read forcing data for all HRUs
  if(simultaneousRead)then
   err=nf90_get_var(ncid,forcFileInfo(iFile)%data_id(ivar),dataVec,start=(/ixHRUfile_min,iRead/),count=(/nHRUlocal,1/))
   if(err/=nf90_noerr)then; message=trim(message)//'problem reading forcing data: '//trim(varName)//'/'//trim(nf90_strerror(err)); return; endif
  endif
```

```fortran
  ! loop through GRUs and HRUs
  do iGRU=1,size(gru_struc)
   do iHRU=1,gru_struc(iGRU)%hruCount
```

```fortran
! get individual data value
if(simultaneousRead) dataVal(1) = dataVec(iHRU_local)
```

```fortran
! put the data into structures
forcStruct%gru(iGRU)%hru(iHRU)%var(ivar) = dataVal(1)
```

Reply:
So: The first code block reads all HRUs (populate `dataVec`) the third code block copies the desired part of `dataVec` into `dataVal` and the fourth code block puts `dataVal` into the data structure.

We could have also written
```fortran
forcStruct%gru(iGRU)%hru(iHRU)%var(ivar) = dataVec(iHRU_local)
```

but I copied `dataVec` into `dataVal` so that I could use the same code to operate on `dataVal` when `simultaneousRead==.false.`